Competition2017/src/competition/subsystems/drive/commands/DriveForDistanceAtHeadingCommand.java

```diff
@@ -0,0 +1,141 @@
+        this.alignmentPid = pidFactory.createPIDManager("Drive at heading alignment", 1/24d, 0, 0, 0, 0.5, -0.5);
+
+    }
+
+    public void setTargetHeading(double heading) {
+        targetHeadingSupplier = () -> heading;
+    }
+    
+    public void setTargetHeadingProp(DoubleProperty heading){
+        targetHeadingSupplier = () -> heading.get();
+    }
+    
+    public void setTargetHeadingSupplier(DoubleSupplier headingSupplier) {
+        targetHeadingSupplier = headingSupplier;
+    }
+
+    public void setTargetDistance(double distance) {
+        targetDistanceSupplier = () -> distance;
+    }
+    
+    public void setTargetDistanceProp(DoubleProperty distance){
+        targetDistanceSupplier = () -> distance.get();
+    }
+    
+    public void setTargetDistanceSupplier(DoubleSupplier distanceSupplier) {
+        targetHeadingSupplier = distanceSupplier;
+    }
+    
+    public void reset(){
+        log.info("Resetting PID");
+        headingDrivePid.reset();
+        distancePid.reset();
+    }
+    
+    @Override
+    public void initialize() {
+        if(targetHeadingSupplier == null) {
+            log.error("DriveForDistanceAtHeading initialized without a heading setpoint!");
+            targetHeading = null;
+            return;
+        }
+        
+        if(targetDistanceSupplier == null) {
+            log.error("DriveForDistanceAtHeading initialized without a distance setpoint!");
+            targetDistance = 0;
+            return;
+        }
+
+        targetHeading = new ContiguousHeading(targetHeadingSupplier.getAsDouble());
+        targetDistance = targetDistanceSupplier.getAsDouble();
+        
+        initialPosition = poseSubsystem.getFieldOrientedTotalDistanceTraveled();
+        
+        log.info("Initializing DriveForDistanceAtHeading with heading " + targetHeading + " and distance " + targetDistance);
+        reset();
+    }
+
+    @Override
+    public void execute() {
+        if(targetHeading != null) {
+            double errorInDegrees = poseSubsystem.getCurrentHeading().difference(targetHeading);
+            double rotationalPower = headingDrivePid.calculate(0, errorInDegrees);
+            
+            
+            XYPair targetRelativePositionalError = poseSubsystem.getFieldOrientedTotalDistanceTraveled()
+                    .add(initialPosition.scale(-1))
+                    .rotate(-targetHeading.getValue());
+            
+            double translationPowerFactor = Math.max(speedHeadingThresh.get() - Math.abs(errorInDegrees), 0) / speedHeadingThresh.get();
+            double translationPower = distancePid.calculate(targetDistance, targetRelativePositionalError.y) * translationPowerFactor;
+            
+            // Using the translation power as a multiplication factor ensures that we only align when we are moving toward the goal
+            double alignmentPower = alignmentPid.calculate(targetRelativePositionalError.x, 0) * Math.abs(translationPower);
```

Comment:
Since alignment is essentially another "rotation", it should share polarity with rotationalPower - right now it looks reversed.